qiskit/visualization/utils.py

```diff
@@ -220,6 +220,39 @@ def get_bit_label(drawer, register, index, qubit=True, layout=None, cregbundle=T
     return bit_label
 
 
+def get_bit_locations(qregs, cregs, qubits, clbits, reverse_bits):
+    """Build and return the bit_locations dict for all 3 drawers
+
+    Args:
+        qregs (list(QuantumRegister)): list of quantum registers
+        cregs (list(ClassicalRegister)): list of classical registers
+        qubits (list(Qubit)): list of quantum bits
+        clbits (list(Clbit)): list of classical bits
+        reverse_bits (bool): reverse order of the bits if True
+
+    Returns:
+        dict: bit location dict in form {bit: {"register": register, "index": index}}
+
+    """
+    # First load register and index info for the cregs and qregs,
+    # then add any bits which don't have registers associated with them.
+    bit_locations = {
+        bit: {"register": register, "index": index}
+        for register in cregs + qregs
+        for index, bit in enumerate(register)
+    }
+    for index, bit in list(enumerate(qubits)) + list(enumerate(clbits)):
+        if bit not in bit_locations:
+            if reverse_bits:
+                if isinstance(bit, Clbit):
+                    index = len(clbits) - index - 1
+                else:
+                    index = len(qubits) - index - 1
+            bit_locations[bit] = {"register": None, "index": index}
+
+    return bit_locations
```

Comment:
Would this become any simpler by using the new `QuantumCircuit.find_bit`?  That handles bits in multiple registers and loose bits, and it's a constant-time lookup per bit (because `QuantumCircuit` stores the necessary data structure).  Either way, we might want to document somewhere appropriate what happens if a bit is in more than one register - the current behaviour looks like it'll get displayed as being in the last defined register.  I don't have a particular preference, so the current behaviour is fine, but we should document it.

For the future, likely we need to completely think how to display which bits are in which registers, but that's out of scope here.

Reply:
Currently the 3 drawers do not have access to the `QuantumCircuit` object. `circuit_drawer` does have access, but it calls `get_layered_instructions` to create a `list(list(DAGOpNode))` which are the layers that will be displayed, and this is what is passed to the drawers. The `qregs, cregs, qubits, clbits` are passed as parameters to the drawers.

It would be easy enough to pass the `QuantumCircuit` object on down to the drawers. It also might make more sense to do the `_bit_locations` inside `get_layered_instructions` and pass the `BitLocations` object to the drawers at that time.

Never thought of the idea of a bit in more than one register. How does one decide??

Reply:
Ah sure, that's no problem then.

Yeah, it's been a fairly big shift recently where `ClassicalRegister` has been completely demoted from its previous role of defining bits.  Now the individual bit objects (`Clbit`) are the primitive type, and `ClassicalRegister` (m.m. quantum) is just an alias to refer to several bits in one go.  It's not something we need to worry about in this PR, but at some point we might want to think about how best to convey that information visually.